Add diagnostic level stdout for total active particle population.

RyanWollaeger opened the pull request:
## Background

* Diagnostic printout for scalar totals can be a useful gauge of correct simulation behavior.

## Description of Changes

+ Include SimTime object in `RadiationStep` task collection, to get ncycle[_out].
+ Add MPI_Reduce to rank 0 of sum of particle number over cells.
+ Add diagnostic_level user input, restricted to 0 or 1 (for now).
+ Update Parthenon submodule to same version currently set in `Artemis`.

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

RyanWollaeger committed:
Add diagnostic level stdout for total active particle population.

+ Include SimTime object in radiation step, to get ncycle[_out].
+ Add MPI_Reduce to rank 0 of sum of particle number over cells.
+ Add diagnostic_level user input, restricted to 0 or 1 (for now).

RyanWollaeger commented:
This is a pretty small change set, but requires an update to Artemis as well (have that change ready once the jaybenne submodule is updated after this is merged).

pdmullen commented on src/jaybenne/population_control.cpp:
```diff
@@ -85,6 +86,34 @@ TaskStatus ControlPopulation(MeshData<Real> *md) {
+          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
+        },
+        Kokkos::Sum<Real>(totag));
+
```
I think we need a Kokkos::fence() here.

RyanWollaeger replied:
Ah - ok - wasn't sure - I know this dips in/out of the kernel. Will add it here and in the other place.

RyanWollaeger replied:
Oh - there is an MPI call right after this - good catch. Added the fence here and moved this whole routine to a function in jaybenne_utils.hh.

pdmullen commented on src/jaybenne/population_control.cpp:
```diff
@@ -152,6 +181,35 @@ TaskStatus ControlPopulation(MeshData<Real> *md) {
+          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
+        },
+        Kokkos::Sum<Real>(totag));
+
```
and here...

alexrlongne replied:
Are the two added `par_reduce` calls the same such that you could put them in a function?

RyanWollaeger replied:
Good catch - these are essentially identical routines. I was hesitant to do this due to the use of the SparsePack here - but will give it a try.

RyanWollaeger replied:
We can then maybe just reuse this function for energy accounting as well.

alexrlongne replied:
Oh, if the sparse pack makes this hard don't worry about it!

RyanWollaeger replied:
Think have something based on what @pdmullen did for source.cpp - will push shortly.

RyanWollaeger replied:
Just pushed the change that moves this to a function. Following a pattern of a reduction function used in `source.cpp` and defined in `jaybenne_utils.hpp`, the new function is templated on a function argument. In the call now here, this argument is lambda function in which the sparse pack is captured by reference.

pdmullen reviewed (commented):
LGTM! Wondering if you favor this approach as opposed to a history variable?

RyanWollaeger commented:
> LGTM! Wondering if you favor this approach as opposed to a history variable?

Would this be a scalar Parthenon IO? That seems like a good idea - we could certainly route this data there.

RyanWollaeger committed:
Incorporate some review suggestions from A. Long and P. Mullen.

+ Add Kokkos::fence() after par_reduce and before MPI reduce.
+ Refactor global sum reduction into templated function.
+ Wrap jaybenne_utils functions in jaybenne namespace.

RyanWollaeger commented on src/jaybenne/jaybenne_utils.hpp:
```diff
@@ -41,11 +43,40 @@ par_reduce_inner(team_mbr_t team_member, const int kl, const int ku, const int j
       reduction);
 }
 
+template <typename Function, typename U>
```
Here is the new function - found the nifty `MPITypeMap` in Parthenon to use for the MPI reduction step below. Guess we could do something similar with the `MPI_SUM` (e.g. abstract to `MPIOpMap<OP<U>>`, where `OP` is another template parameter for this function).

RyanWollaeger commented:
The latest commit doesn't work on GPU - looking into it (converting the lambda to a KOKKOS_LAMBDA gets the build to work but causes an invalid space access at runtime).

adamdempsey90 commented on src/jaybenne/jaybenne_utils.hpp:
```diff
@@ -41,11 +43,40 @@ par_reduce_inner(team_mbr_t team_member, const int kl, const int ku, const int j
+      jl, ju, il, iu,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, U &tota) {
+        function(b, k, j, i, tota);
+      },
```
```suggestion
      function,
```

RyanWollaeger replied:
Done

adamdempsey90 commented on src/jaybenne/population_control.cpp:
```diff
@@ -85,6 +87,24 @@ TaskStatus ControlPopulation(MeshData<Real> *md) {
+        "ControlPopulation::report-tot-active-1",
+        [&](const int &b, const int &k, const int &j, const int &i, Real &tota) {
+          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
+        },
```
```suggestion
       KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i, Real &tota) {
          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
        },
```
as you said in your comment. You've basically made an abstraction of the parthenon abstraction, so you would follow the same logic `par_reduce` does with respect to the kokkos `parallel_reduce`. Because of that, I'd also suggest moving the label to the first argument and passing the execution space here as the second argument.

RyanWollaeger replied:
Done

RyanWollaeger replied:
Oh wait - forgot to sort the arguments and add the execution space.

RyanWollaeger replied:
Now the label is reordered (and actually used), and the `ExecSpace` is now a templated (2nd) argument to the function.

adamdempsey90 commented on src/jaybenne/population_control.cpp:
```diff
@@ -152,6 +172,24 @@ TaskStatus ControlPopulation(MeshData<Real> *md) {
+        "ControlPopulation::report-tot-active-2",
+        [&](const int &b, const int &k, const int &j, const int &i, Real &tota) {
+          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
+        },
```
```suggestion
        KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i, Real &tota) {
          tota += vmesh(b, fj::active_num_per_cell(), k, j, i);
        },
```

adamdempsey90 commented on src/jaybenne/jaybenne_utils.hpp:
```diff
@@ -41,11 +43,40 @@ par_reduce_inner(team_mbr_t team_member, const int kl, const int ku, const int j
+template <typename Function, typename U>
+void global_sum_reduce(const int nblocks, const int kl, const int ku, const int jl,
+                       const int ju, const int il, const int iu, const std::string label,
+                       const Function &function, U &globally_reduced) {
```
This might be useful enough to move into parthenon proper

RyanWollaeger replied:
Good call on these changes - will try them. Think something else might be wrong with my configuration when I was testing on Darwin GPU, which I wasn't encountering on Chicoma - the latter of my errors (segfault) didn't go away when I reverted this commit...

RyanWollaeger replied:
Fixed my issue - this is working now with your suggested changes - getting correct `stepdiff` pop. data off the GPU, and on 2 MPI ranks.

RyanWollaeger replied:
Made the changes here - also haven't yet abstracted the MPI/Kokkos reduction operation (using the aforementioned map approach) - but we could open a PR there as well.

adamdempsey90 replied:
For Jaybenne, I wouldn't bother abstracting too much. Just have what you need for now. If you want to move this into parthenon, though, then more abstractions would be good (and things like passing multiple reducers).

RyanWollaeger committed:
Incorporate A. Dempsey's GPU fix for new function.

+ Pass in KOKKOS_LAMBDA as function-template function argument.
+ Simplify function usage in the new function.

RyanWollaeger committed:
Address other feedback from A. Dempsey.

+ Add ExecSpace template argument to global reduction function.
+ Make label argument first (also use it).

RyanWollaeger commented:
> > LGTM! Wondering if you favor this approach as opposed to a history variable?
> 
> Would this be a scalar Parthenon IO? That seems like a good idea - we could certainly route this data there.

Having read some of the history code in Parthenon now, it seems this is almost functionally equivalent - in fact less general in reductions as-written. So sort of reinvented the wheel here, unless if one wants this global operation not tied to history IO, maybe (but I'm not sure when that would necessarily be true).

RyanWollaeger commented:
Maybe the thing to do is keep this for now (this is a fairly small function), and eventually tie history-based output to the `diagnostic_level` as an additional option.

pdmullen commented:
LGTM! Feel free to merge when ready :)

RyanWollaeger commented:
Will merge and then update artemis - thanks for the great reviews!
